### PR TITLE
feat(6178): emit ruby enum as integer

### LINKED ISSUE #6178: Ruby JSON serialize enums as ints
**What language does this apply to?**
Ruby proto3

**Describe the problem you are trying to solve.**
The proto3 docs says that a JSON implementation may provide an option to:
> Emit enum values as integers instead of strings: The name of an enum value is used by default in JSON output. An option may be provided to use the numeric value of the enum value instead.

Currently this isn't implemented by the Ruby library. 

**Describe the solution you'd like**
As far as I know of this feature is implemented by the [C++](https://github.com/protocolbuffers/protobuf/issues/2735) and [Go](https://godoc.org/github.com/golang/protobuf/jsonpb#Marshaler) libraries. I would like to request this so that Ruby has feature parity with the other supported languages.

**Additional context**
I believe the entry point to Ruby's encode_json method is here:
https://github.com/protocolbuffers/protobuf/blob/f425b9f059a57dc5f13a6e6a389fff1a6cc06749/ruby/ext/google/protobuf_c/encode_decode.c#L1477

### PR DESCRIPTION
**Pull request by MQuy**
Fixes #6178

- Add a new option `enums_as_integers` to emit enum as integer value.

### CODE REVIEW

File: ruby/tests/encode_decode_test.rb [hunk 1]
```diff
@@ -84,6 +84,34 @@ def test_encode_json
     )
 
     assert_match 'optional_int32', json
+
+
+    # Test for enums printing as ints.
```
Comment (MQuy) [inline ruby/tests/encode_decode_test.rb:89]: thank @spiffyy99 for the tests https://github.com/protocolbuffers/protobuf/pull/6384

**Comment by JasonLunn** [comment]
Thanks for the contribution, @MQuy. Can you add the corresponding changes to the JRuby implementation?

**Comment by quy-cbh** [comment]
hi @JasonLunn, I added JRuby implementation for emitting enums as integers. It looks like the pipeline won't run except it is triggered from your side?

**Comment by MQuy** [comment]
hi @JasonLunn, I added JRuby implementation for emitting enums as integers. It looks like the pipeline won't run except it is triggered from your side?

**Comment by haberman** [comment]
This PR looks good. Would you be able to rebase it on main? Sorry for the trouble, you've caught us in the middle of a migration to GitHub Actions.

File: ruby/ext/google/protobuf_c/message.c [hunk 2]
```diff
@@ -1171,6 +1171,11 @@ static VALUE Message_encode_json(int argc, VALUE* argv, VALUE klass) {
                               Qfalse))) {
       options |= upb_JsonEncode_EmitDefaults;
     }
+
+    if (RTEST(rb_hash_lookup2(hash_args, ID2SYM(rb_intern("enums_as_integers")),
```
Comment (haberman) [inline ruby/ext/google/protobuf_c/message.c:1175]: This name should be `format_enums_as_integers`, per our internal document on this new API.

Comment (MQuy) [reply ruby/ext/google/protobuf_c/message.c]: thank you for the suggestion, updated.

**Comment by haberman** [comment]
I'm seeing errors in the tests:

> error: cannot find symbol
>      if (enumAsIntegers != null && printingEnumsAsInts.isTrue()) ***

**Comment by quy-cbh** [comment]
> I'm seeing errors in the tests:
> 
> > error: cannot find symbol
> > if (enumAsIntegers != null && printingEnumsAsInts.isTrue()) ***

@haberman, my bad, updated

**Comment by MQuy** [comment]
hi @haberman, @zhangskz and @JasonLunn, would love to have your updates on this feature.

**Review by haberman — APPROVED** [review]

**Comment by ericsalo** [comment]
I do apologize for having to ask this of you again but could you please rebase one more time? Things appear to be stuck even though the PR has been approved. Thanks.

**Comment by MQuy** [comment]
@ericsalo no worry, I rebased. Can you trigger pipelines again and merge the pull request. Thank you

**Review by ericsalo — APPROVED** [review]

**Comment by MQuy** [comment]
hi @haberman, @zhangskz, @JasonLunn and @ericsalo is it possible that we include this feature in the next release?

**Pull request closed without merging**